src/app/core/services/group-watching.service.ts

```diff
@@ -0,0 +1,108 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { ActivatedRoute, Params, Router } from '@angular/router';
+import { BehaviorSubject, Observable, of } from 'rxjs';
+import { distinctUntilChanged, filter, map, shareReplay, startWith, switchMap, withLatestFrom } from 'rxjs/operators';
+import { GetGroupByIdService, Group } from 'src/app/modules/group/http-services/get-group-by-id.service';
+import { GetUserService, User } from 'src/app/modules/group/http-services/get-user.service';
+import { isNotUndefined } from 'src/app/shared/helpers/null-undefined-predicates';
+import { rawGroupRoute, RawGroupRoute } from 'src/app/shared/routing/group-route';
+
+const watchedGroupQueryParam = 'watchedGroupId';
+const watchedGroupIsUserQueryParam = 'watchUser';
+const noWatchingValue = 'none'; // value of `watchedGroupQueryParam` which means that there is no watching
```

Comment:
The query params are weird to me because you can have unexpected combinations like `watchGroupId=none&watchUser=[whatever]`, `watchGroupId=[actualGroupId]&watchUser=toto`
Then based on a special value we will unwatch by returning `null` instead of `undefined`, meaning we end up with a distinction between `null` and `undefined`, looks smelly to me :grimacing: 

If I understand it correctly, what we want to do is:
- Be able to watch a group or a user by user navigation (a button) or url navigation (`watchedGroupId`)
- Be able stop watching by user navigation (a button) or url navigation (for dev/admin usage only I suppose ?)
- For a technical reason, be able to make the differentiate a watched group and a watched user

To me the query parameters handling should be more something like:
1. A query param `watchedGroupId=…` (like now) and check existence of param `watchedGroupIsUser` (for instance) instead of allowing any value as falsy
2. Two query params `watchedGroupId` and `watchedUserId`, but then we'd need to throw an error if both are supplied

As for the special value `'none'`, it could be replaced by the special case `watchedGroupId=` meaning the a watched group id exists but it's empty, or another query param `stopWatching`, but it induces to throw an error as well if both query params are supplied.

Thoughts ?

Reply:
> The query params are weird to me because you can have unexpected combinations like `watchGroupId=none&watchUser=[whatever]`, `watchGroupId=[actualGroupId]&watchUser=toto` Then based on a special value we will unwatch by returning `null` instead of `undefined`, meaning we end up with a distinction between `null` and `undefined`, looks smelly to me 😬

There are different aspects here.

1. IMO, making a distinction between `null` and `undefined` is not smelly at all, there are two values in JS, not one, "null" meaning there is explicitly nothing, `undefined` meaning the value is not-defined/unknown. The semantic between them is quite clear to me (even if sometimes wrongly used in the code).

2. That's because I prefer using '1'  or '0' in the url than using an empty parameter in the url to mean "true". IMO "mypage?is_user=1&something=434" is clearer than "mypage?is_user=&something=434" to mean "is_user=true" and something=434 meaning false... especially as this whole PR make the assumption (because of the "angular limitation") that "no arg given" means "unchanged"  I saw in the meantime in LTI that you use the empty parameter in LTI. The problem is that query strings are strings and we want boolean so we have to bind an infinite set of strings to two values. So the expected values are simply "0" or "1"... the other (if somebody manually changes the url) are just interpreted as "0". In your case, "mypage?is_user=0&something=434" would mean "is_user=true" which one could find strange as well.

3. We can have inconsistent parameter set "like `watchGroupId=none&watchUser=[whatever]`", right. But I don't see anyway to prevent them, if having a single parameter `watchGroupId` or `watchUserId`, we have the same inconsistency problem if both are given. These are inputs, given only as strings, so as any input they have to be sanitize somehow to prevent impossible behavior as any set of query strings. Of course, I could document a bit more what parameter has the priority over the other one if we think it is useful (as any piece of code could navigate to such url).

> If I understand it correctly, what we want to do is:
> 
>     * Be able to watch a group or a user by user navigation (a button) or url navigation (`watchedGroupId`)
> 
>     * Be able stop watching by user navigation (a button) or url navigation (for dev/admin usage only I suppose ?)

"for dev/admin usage only I suppose ?" -> From what I remember from my code, not really, the change of "watching" is only driven by url (to have a single "entry point" for changing it)... So every change is driven by url change... there is just (because of the "angular limitation") that if params are not given, we consider the previous situation as unchanged.

> 
>     * For a technical reason, be able to make the differentiate a watched group and a watched user
> 
> 
> To me the query parameters handling should be more something like:
> 
>     1. A query param `watchedGroupId=…` (like now) and check existence of param `watchedGroupIsUser` (for instance) instead of allowing any value as falsy
> 
>     2. Two query params `watchedGroupId` and `watchedUserId`, but then we'd need to throw an error if both are supplied
> 

So these are two separate options, right?  Option 2 does not solve any inconsistency as we have to decide what happens in I get "mypage?watchedGroupId=1&watchedUserId=2". This option was considered and I preferred the other one, so coming back to your option 1.

Regarding your option 1, as I explain in (2) above, that is a choice to be done. It is just that it is designed to be "=0" or "=1. As explained above, the main problem I have with your option is that a parameter not set means "unchanged" in group watching, making it inconsistent with how `watchedGroupId` works.
 
> As for the special value `'none'`, it could be replaced by the special case `watchedGroupId=` meaning the a watched group id exists but it's empty

I could do that.

> or another query param `stopWatching`, but it induces to throw an error as well if both query params are supplied.

Yes, of course it was also consider but it involves more confusion.

So, maybe we should do quick chat to pick a choice for all these.